File: 1721-swapping-nodes-in-a-linked-list/1721-swapping-nodes-in-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        ListNode fast = head, slow = head, k_from_st = null;
        int i = k - 1;
        
        while (fast != null && i-- > 0){
            fast = fast.next;
        }
        
        k_from_st = fast;
        fast = fast.next;
        while (fast != null){
            fast = fast.next;
            slow = slow.next;
        }        
        
        int temp = slow.val;
        slow.val = k_from_st.val;
        k_from_st.val = temp;
        
        return head;
    }
}

File: 1721-swapping-nodes-in-a-linked-list/1721-swapping-nodes-in-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        lead = lag = dummy
        
        for i in range(k + 1):
            lead = lead.next
            
        while lead:
            lead = lead.next
            lag = lag.next
           
        cur = dummy
        for i in range(k - 1):
            cur = cur.next
        
        tmp = cur.next.val
        cur.next.val = lag.next.val
        lag.next.val = tmp
        
        return dummy.next
        
            
        
        
        
        